src/Database/PostgreSQL/PQTypes/Internal/State.hs

```diff
@@ -1,49 +1,273 @@
+          _ -> "ROLLBACK"
+    _ <- liftBase . uninterruptibleMask_ $ runQueryIO @SQL conn finalizeSql
+    putConnection ics (conn, cdata) ec
+  Finalized -> error "finalized connection"
+
+----------------------------------------
+
+data ConnectionData m = forall cdata. ConnectionData
+  { cdConnectionSource :: !(InternalConnectionSource m cdata)
+  , cdConnectionState :: !(MVar (ConnectionState cdata))
+  }
+
+getConnectionSource :: ConnectionData m -> ConnectionSourceM m
+getConnectionSource ConnectionData {..} = ConnectionSourceM cdConnectionSource
+
+getConnectionAcquisitionModeIO
+  :: HasCallStack
+  => ConnectionData m
+  -> IO ConnectionAcquisitionMode
+getConnectionAcquisitionModeIO ConnectionData {..} = do
+  readMVar cdConnectionState >>= \case
+    OnDemand -> pure AcquireOnDemand
+    Acquired isolationLevel permissions _ _ -> do
+      pure $ AcquireAndHold isolationLevel permissions
+    Finalized -> error "finalized connectio"
+
+withConnectionData
+  :: (HasCallStack, MonadBase IO m, MonadMask m)
+  => ConnectionSourceM m
+  -> TransactionSettings
+  -> (ConnectionData m -> m r)
+  -> m r
+withConnectionData cs ts action = (`fix` 1) $ \loop n -> do
+  let maybeRestart = case tsRestartPredicate ts of
+        Just _ -> handleJust (expred n) $ \_ -> loop $ n + 1
+        Nothing -> id
+  maybeRestart
+    . fmap fst
+    . generalBracket (initConnectionData cs cam) finalizeConnectionData
+    $ action
+  where
+    cam = tsConnectionAcquisitionMode ts
+
+    expred :: Integer -> SomeException -> Maybe ()
+    expred n e = do
+      -- check if the predicate exists
+      RestartPredicate f <- tsRestartPredicate ts
+      -- cast exception to the type expected by the predicate
+      err <-
+        msum
+          [ -- either cast the exception itself...
+            fromException e
+          , -- ...or extract it from DBException
+            fromException e >>= \DBException {..} -> cast dbeError
+          ]
+      -- check if the predicate allows for the restart
+      guard $ f err n
+
+changeAcquisitionModeTo
+  :: (HasCallStack, MonadBase IO m, MonadMask m)
+  => ConnectionAcquisitionMode
+  -> ConnectionData m
+  -> m ()
+changeAcquisitionModeTo cam ConnectionData {..} = mask_ $ do
+  bracketOnError (takeMVar cdConnectionState) (putMVar cdConnectionState) $ \case
+    OnDemand -> case cam of
+      AcquireOnDemand -> putMVar cdConnectionState OnDemand
+      _ -> do
+        newConnState <- initConnectionState cdConnectionSource cam
+        putMVar cdConnectionState newConnState
+    connState@(Acquired isolationLevel permissions _ _) -> case cam of
+      AcquireOnDemand -> do
+        finalizeConnectionState cdConnectionSource (ExitCaseSuccess ()) connState
+        putMVar cdConnectionState OnDemand
+      AcquireAndHold newIsolationLevel newPermissions -> do
+        when (isolationLevel /= newIsolationLevel) $ do
+          error $
+            "isolation level mismatch (current: "
```

Comment:
I think we could allow lower isolation levels, no?

Reply:
In principle, but it's simpler to only allow the same level, I don't really want to think about potential consequences of allowing this, especially since we don't use it.